### ISSUE TITLE
SSH cmd sending and reciving

### ISSUE THREAD

**Issue by Clio75** [fieldrndservices/libssh2-labview#3]
Hi, 

I'm trying to use this toolkit to send and received cmd over ssh. I'm find the Execute that sending CMD. but how to received the answer ? 

Great toolkit  :D Looking forward to test it more

**Comment by volks73** [comment]
Since libssh2-labview is a wrapper around the [libssh2 C library](https://www.libssh2.org), but with a [C "adapter" library](https://github.com/fieldrndservices/libssh2lv) to make the libssh2 library more LabVIEW-friendly, the [examples](https://www.libssh2.org/examples/) for the libssh2 library can be helpful. There is the [ssh2_echo.c](https://www.libssh2.org/examples/ssh2_echo.html) and [ssh2_exec.c](https://www.libssh2.org/examples/ssh2_exec.html) examples that can act as a guide for reading the response from a command over SSH. However, I recognize the examples are a little low-level, and written in C, which is sometimes difficult to reason about and/or maybe unfamiliar.

Unfortunately, I do not currently have an example in LabVIEW for each of the libssh2 C library examples, which would be the most helpful. The current examples focus on SFTP, which was the main functionality needed by another collaborate at the initiation of this project. I will create a feature request/issue to add more examples.

There is the Channel.lvlcass:Read All.vi from the Channel "class". This is probably the VI you want to use after executing the command. It encapsulates a lot of the reading functionality that appears in the C echo and execute examples. Here is a totally untested VI snippet that might work:

![Example Execute Read](https://user-images.githubusercontent.com/1915469/77656696-a08bc900-6f4a-11ea-943d-cc2628e394d0.png)

**Comment by volks73** [comment]
I have created #4 to track creation of examples.

Please let me know if the above VI snippet works and would be a good example.

**Comment by Clio75** [comment]
I will test and give you feedback shortly :)

**Comment by Clio75** [comment]
Hi, 

I have one problem
![image](https://user-images.githubusercontent.com/8381796/77668112-49eeb100-6f83-11ea-9360-fe53453bf7ba.png)

Red square : Do I have correct destroy channel ? 
I can't find the same symbol as you have in your snippet.

Yellow square : Is this correct ? 

Let me know what I'm doing wrong :)

**Comment by volks73** [comment]
The Close.vi should be Destroy.vi. In the toolkit, for every Create there is a Destroy and for every Open (sometimes buried within a Create or Initialize VI) there is a Close VI. The wrong VI appeared to be used in this palette. Please see the attached ZIP/VIP with the updated palette that contains the missing Destroy.vi.

However, my example should be amended to include using the Close.vi in addition to the Destroy.vi. Here is an updated VI snippet, still untested, and might work:

![Channel Exec-Read Example v2](https://user-images.githubusercontent.com/1915469/77675201-008e6980-6f63-11ea-9e60-f5122080eb09.png)

[field_rnd_services_libssh2-1.0.0.9.zip](https://github.com/fieldrndservices/libssh2-labview/files/4388647/field_rnd_services_libssh2-1.0.0.9.zip)

**Comment by Clio75** [comment]
Perfect 💃 🥇

**Comment by Clio75** [comment]
Just a comment: 
Can the close channel be inside the destroy channel. I believe that is more intuitive for the user 
Like you also sad, Create -> destroy. Open-> Close.

**Comment by volks73** [comment]
>Can the close channel be inside the destroy channel. I believe that is more intuitive for the user
Like you also sad, Create -> destroy. Open-> Close.

I agree, and I have a vague reconcolation not really like this inconsistency within the API for the channels, but there appears to be some occasions where the channel does not always need to be closed, but it does always need to be destroyed, a.k.a. freed. Looking at the examples for the libssh2 library, only the [ssh2.c](https://www.libssh2.org/examples/ssh2.html), [ssh2_agent.c](https://www.libssh2.org/examples/ssh2_agent.html), [ssh2_echo](https://www.libssh2.org/examples/ssh2_echo.html), and [ssh2_exec.c](https://www.libssh2.org/examples/ssh2_exec.html) examples explicitly use the [`libssh2_channel_close`](https://www.libssh2.org/libssh2_channel_close.html) function (Close.vi), but all examples use the [`libssh2_channel_free`](https://www.libssh2.org/libssh2_channel_free.html) function (Destroy.vi). 

All of the examples use a variant of the [`libssh2_channel_open_ex`](https://www.libssh2.org/libssh2_channel_open_ex.html) function to create the channel, so the Create.vi is actually polymorphic and the type of open function is determined by the instance of the Create.vi selected. The default instance is to use the `libssh2_channel_open_session` function. I remember combining the creation of the channel with the opening of a channel as a convenience and to decrease the number of VIs needed to "get going", and opening was always required.

A possible complication, if you look at the [VI snippet in the README for uploading a file to a remote SSH server](https://raw.githubusercontent.com/fieldrndservices/libssh2-labview/master/docs/SCP-File-Upload.png), there is the Close Channel and Wait on Close Channel VIs. So, how would including the Close VI in with the Destroy VI work if the user of the toolkit wanted to wait for the close signal to be received before destroying the channel?

Basically, the inconsistency is a result of the libssh2 implementation and trying to create an ergonomic wrapper between C and LabVIEW. There is an Open and Close functions within the libssh2 library, but the Open is "morphed" into the Create functionality for the adapter library and the LabVIEW toolkit. 

I think technically the Close is not actually needed, but good practice. I also think that always executing a Close when the channel is destroyed may not hurt as well.

This is all probably way more information than you want or need, but I am going to create a separate issue to track resolution of this inconsistency, which does annoy me, too.

**Comment by volks73** [comment]
See Issue #6 for tracking the inconsistent naming. I believe this issue has been resolved.
